src/control/server/server.go

```diff
@@ -424,9 +424,13 @@ func Start(log *logging.LeveledLogger, cfg *config.Server) error {
 			switch evt.ID {
 			case events.RASSwimRankDead:
 				// Mark the rank as unavailable for membership in
-				// new pools, etc.
-				if err := membership.MarkRankDead(system.Rank(evt.Rank)); err != nil {
-					log.Errorf("failed to mark rank %d as dead: %s", evt.Rank, err)
+				// new pools, etc. but ignore events for already
+				// evicted ranks.
+				err := membership.MarkRankDead(system.Rank(evt.Rank))
+				if err != nil {
+					if errors.Cause(err) != system.ErrRankAlreadyEvicted {
```

Comment:
Hmm. This is fine, but I wonder if it would be cleaner to avoid exporting the error outside of the system package and instead just check for the unexported sentinel error in `membership.MarkRankDead` and return nil in that case. That keeps the logic for ignoring this error contained in one place. If we ever need to mark ranks dead in other parts of the code, we'd have to duplicate this logic there too.

Reply:
maybe I'm misunderstanding but if we do what you suggest, won't we do the group update when transitioning from evicted to evicted which is exactly what we wanted to avoid in the first place?

Reply:
Ah, right. Of course. Hrm. I think the thing I'm having a bit of trouble with is the new exported sentinel error. We're adding this new thing so that we can see from outside of the method that we failed to mark a rank as dead, but the only thing we're doing in that case is logging it and then returning early.

Given that there isn't really any useful error handling that can be done in this case, I think I would still prefer to avoid exporting the error outside of the system package. We can still log the failure (in cases other than evicted->evicted) in `membership.MarkRankDead()` and return an error, and then out here in the server package we just return early when `err != nil`.